libraries/libfc/include/fc/crypto/bls_signature.hpp

```diff
@@ -1,51 +1,136 @@
+         constexpr bool raw   = false; // non-montgomery
+         std::optional<bls12_381::g2> g2 = bls12_381::g2::fromAffineBytesLE(sig._affine_non_montgomery_le, check, raw);
+         FC_ASSERT(g2, "Invalid bls signature ${k}", ("k", sig._affine_non_montgomery_le));
+         sig._jacobian_montgomery_le = *g2;
+         return ds;
+      }
+
+   private:
+      std::array<uint8_t, 192> _affine_non_montgomery_le{};
+      bls12_381::g2            _jacobian_montgomery_le; // cached g2
+   };
 
-         bls_signature& operator= (const bls_signature& ) = default;
+   // Serialization form:
+   //   Non-Montgomery form and little-endian encoding for the field elements.
+   //   Affine form for the group element (the z component is 1 and not included in the serialization).
+   //   Binary serialization encodes x component first followed by y component.
+   class bls_aggregate_signature {
+   public:
+      bls_aggregate_signature() = default;
+      bls_aggregate_signature(bls_aggregate_signature&&) = default;
+      bls_aggregate_signature(const bls_aggregate_signature&) = default;
+      bls_aggregate_signature& operator=(const bls_aggregate_signature&) = default;
+      bls_aggregate_signature& operator=(bls_aggregate_signature&&) = default;
 
-         // affine non-montgomery base64url with bls_signature_prefix
-         std::string to_string(const yield_function_t& yield = yield_function_t()) const;
+      // affine non-montgomery base64url with bls_signature_prefix
+      explicit bls_aggregate_signature(const std::string& base64urlstr);
 
-         bool equal( const bls_signature& sig ) const;
+      explicit bls_aggregate_signature(const bls_signature& sig)
+         : _jacobian_montgomery_le(sig.jacobian_montgomery_le()) {}
 
-         bls12_381::g2 _sig;
+      // aggregate signature into this
+      void aggregate(const bls_signature& sig) {
+         _jacobian_montgomery_le.addAssign(sig.jacobian_montgomery_le());
+      }
+      // aggregate signature into this
+      void aggregate(const bls_aggregate_signature& sig) {
+         _jacobian_montgomery_le.addAssign(sig.jacobian_montgomery_le());
+      }
 
-   }; // bls_signature
+      // affine non-montgomery base64url with bls_signature_prefix
+      std::string to_string() const;
+
+      const bls12_381::g2& jacobian_montgomery_le() const { return _jacobian_montgomery_le; }
+
+      bool equal( const bls_aggregate_signature& sig) const {
+         return _jacobian_montgomery_le.equal(sig._jacobian_montgomery_le);
+      }
+
+      template<typename T>
+      friend T& operator<<(T& ds, const bls_aggregate_signature& sig) {
+         constexpr bool raw = false;
+         std::array<uint8_t, 192> affine_non_montgomery_le = sig._jacobian_montgomery_le.toAffineBytesLE(raw);
+         ds.write(reinterpret_cast<const char*>(affine_non_montgomery_le.data()), affine_non_montgomery_le.size()*sizeof(uint8_t));
+         return ds;

```

Comment:
Is this OK? I though there was an issue with the `jacobian_montgomery` form being non-unique? Would a serialization + deserialization get us back to the same value?

Reply:
@arhag ?

Reply:
Serialization then deserialization would not necessarily give back the original value. But that is not a problem